lib/mllp/client.ex

```diff
@@ -356,282 +381,384 @@ defmodule MLLP.Client do
+
+  defp unexpected_message(state, event, message) do
+    Logger.warn(
+      "Event: #{inspect(event)} in state #{state}. Unknown message received => #{inspect(message)}"
+    )
+
+    :keep_state_and_data
+  end
+
+  ## Handle the (fragmented) responses to `send` request from a caller
+
+  defp handle_received(_reply, %{caller: nil} = data) do
+    ## No caller, ignore
+    data
+  end
+
+  defp handle_received(<<@header, _rest::binary>> = reply, data) do
+    receive_impl(reply, data)
+  end
+
+  ## No header in the first packet
+  defp handle_received(_reply, %{receive_buffer: []} = data) do
+    reply_to_caller({:error, :invalid_reply}, data)
+  end
+
+  ## The rest of MLLP (after the header was received)
+  defp handle_received(reply, data) do
+    receive_impl(reply, data)
+  end
+
+  def receive_impl(reply, %{receive_buffer: buffer, last_byte_received: last_byte} = data) do
+    new_buf = update_receive_buffer(buffer, reply)
+
+    case trailer_check(reply, last_byte) do
+      :data_after_trailer ->
+        Logger.error("Client #{inspect(self())} received data following the trailer")
+        reply_to_caller({:error, :data_after_trailer}, data)
+
+      true ->
+        Logger.debug("Client #{inspect(self())} received a full MLLP!")
+        reply_to_caller({:ok, IO.iodata_to_binary(new_buf)}, data)
+
+      false ->
+        Logger.debug("Client #{inspect(self())} received a MLLP fragment: #{reply}")
+
+        data
+        |> Map.put(:receive_buffer, new_buf)
+        |> Map.put(:last_byte_received, String.last(reply))
     end
   end
 
-  defp do_recv(state, length, :infinity) do
-    res = state.tcp.recv(state.socket, length, :infinity)
-    {res, 0}
+  ## "Split trailer" case
+  ##
+
+  defp trailer_check(<<@cr, rest::binary>>, @eb) do
+    if rest == "" do
+      true
+    else
+      :data_after_trailer
+    end
   end
 
-  defp do_recv(state, length, timeout) do
-    timeout_in_ms = System.convert_time_unit(timeout, :microsecond, :millisecond)
-    t1 = System.monotonic_time(:microsecond)
-    res = state.tcp.recv(state.socket, length, timeout_in_ms)
-    t2 = System.monotonic_time(:microsecond)
-    {res, t2 - t1}
+  defp trailer_check(packet, _last_buffer_byte) do
+    case :binary.match(packet, @trailer) do
+      :nomatch -> false
+      {pos, @trailer_length} when pos + @trailer_length == byte_size(packet) -> true
+      _ -> :data_after_trailer
+    end
   end
 
-  defp update_recv_time_credit(:infinity, _), do: {:infinity, 0}
+  defp update_receive_buffer(buffer, packet) do
+    case buffer do
+      [] ->
+        [packet]
 
-  defp update_recv_time_credit(time_left, time_spent) do
-    time_charged = div(time_spent, 1000) * 1000
-    time_owed = time_spent - time_charged
-    {time_left - time_charged, time_owed}
+      b ->
+        [b | packet]
+    end
+  end
+
+  defp reply_to_caller(reply, %{caller: caller, context: context} = data) do
+    caller && :gen_statem.reply(caller, format_reply(reply, context))
```

Comment:
I think we can avoid the boolean logic here, in fact if we're ending up here when caller is `nil`, we have a bug somewhere further up and it would be best to crash.

Reply:
This is to avoid situations where the receiver sends more packets after full MLLP has been received.
It would not be a bug on our side, but rather a receiver misbehaving. Do you think we should crash in this case?

Reply:
Seems like we'd hit the clause at line 577, no? This raises an interesting question none the less. Crashing aside, one place a `RST` is appropriate is when the the remote sends data that is unexpected. I wonder if we should shutdown in this case? Feels heavy handed, but yeah 🤔

Reply:
> Seems like we'd hit the clause at line 577, no? 

You are correct, we are probably covered for the "more packets" case. Although I'm not entirely sure, as there could be a race condition with the timer in the "receiving" state. There is actually a more serious reason for checking if we have a caller - the receiver can disconnect at any given time. If it happens while the 'send' call is in progress, we notify the caller, otherwise, we don't (no caller there).

If we want to shut down on the "more packets", the natural place is around line 577 indeed.

Reply:
> Although I'm not entirely sure, as there could be a race condition with the timer in the "receiving" state.

Not sure, will have to stare a bit more and we'll both have to attempt to put it through the ringer :) 

I think shutting down makes sense, I think we can just use `maybe_close` , we might need to make a modification though (i.e., maybe sometimes we want to do a graceful close and maybe some times an abortive close).

 Right now maybe_close will attempt to do a graceful close, in this case it's not what we want. We want `:gen_tcp.shutdown(socket, :read_write)`. This is the same behavior as `send_timeout_close: true`, which ends up resulting in a `sock_select()` call that does an abortive close (`RST`).  This should be double checked though. 

Speaking of `send_timeout_close: true`, if we don't have it this set as the default in sock opts we should, the only reason it's false in otp is for backwards compat reasons (though it should be configurable by the user).

Reply:
the send_timout_close bit was resolved by your last commit @bokner. WE still need to hang up the phone on `receive_timeout` as well. Basically, per the nature of MLLP and the implicit behaviour when shipping HL7 around, there's may be no recourse in a situation where the receive timed out. 

Specifically, the client can either hang up the phone or make assumptions about how the server is implemented. In this regard, both the client and server may end up in a bad state and form an error loop. One example where a receive timeout can result in  bad things happening is you peel off part of the response, you time out waiting for the rest. The caller of the client decides to send again, yet there still data in transit (i.e., not in the local buffers and bubbled up to our app yet), the send happens, and now the rest of the ack from a previous response arrives. Meanwhile, the remote might get what you just sent and either accept it, and send back a nack or hang up the phone. 

Tricky business 😁 

Thoughts?

Reply:
The downside of dropping the connection on the client side for whatever reason (say, `receive_timeout`) is that then it would have to reconnect again. Maybe it is a better recourse after all.

I guess the sane client would try to check for connection on any error anyway.

Reply:
Yeah, and to add to that, there needs to be options for all of it, that really bothers me, I think having too many options is never great, but sadly since MLLP isn't a much of a protocol, the behaviour is going to differ from system to system. All we can do is try to provide a mostly sane set of defaults that cover most cases. If I go by the "protocol" and how tcp works, it leads me here to these decisions.

Reply:
I tell ya what, let's not agonize over this right now. We're not at 1.0, code is easy to change, etc. Let's get test passing, merge, and take it from there.